### PR TITLE
Get Run Control value for Web dashboard

### PR DESCRIPTION
**Pull request by morgandavies1997**
Relates to https://github.com/ISISComputingGroup/IBEX/issues/2070

Included webserver tests but may need refractoring to test actually production code

To test the code, we should expect a cross or a tick if the run control values are in range or not

### CODE REVIEW

File: display_blocks.js [hunk 1]
```diff
@@ -116,7 +116,7 @@ function clear(node) {
  */
 function refresh() {
 	$.ajax({
-		url: "http://dataweb.isis.rl.ac.uk:" + PORT + "/",
```
Comment (mattclarke) [inline display_blocks.js]: Was this change to localhost meant to be checked in?

File: get_block_details.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        return title
+
+
+def get_rc_values_for_block(block_name, pvs):
+    """Search pvs for RC values for given block and return them"""
+    for k, v in pvs.items():
+        block_names = k.split(':')
+        if block_name == block_names[0]:
+            print block_name, v
+    return v
```
Comment (samueljackson92) [inline get_block_details.py]: This always returns the last value (`v`) in the dict of `pvs`. Worse still, this assumes dictionaries are ordered which is *sort of* true but not guaranteed in 2.7 and it completely not true in 3+. I assume we actually want to break as soon as we find a `block_name` that matches.

File: get_block_details.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+
+def set_rc_values_for_block_from_pvs(block_object, pvs):
+    """Search pvs for RC values for given block and return them"""
+    block_name = block_object.get_name()
+    items = pvs.items()
+
+    for k, v in items:
+        test_block_name = k.split(':')[0]
+        if block_name == test_block_name:
```
Comment (samueljackson92) [inline get_block_details.py]: Could use `continue` to reduce complexity

File: get_block_details.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+            if "LOW.VAL" == k.split(':')[-1]:
+                low_rc = block_object.set_rc_low(v.get_value())
+            if "HIGH.VAL" == k.split(':')[-1]:
+                high_rc = block_object.set_rc_high(v.get_value())
+            if "INRANGE.VAL" == k.split(':')[-1]:
+                inrange_rc = block_object.set_rc_inrange(v.get_value())
+                return low_rc, high_rc, inrange_rc
```
Comment (samueljackson92) [inline get_block_details.py]: Return should be un-indented here and the `*_rc`s should be initialized above.

File: get_block_details.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+
+def set_rc_values_for_blocks(block_objects, pvs):
+    """Set all RC values for all the given blocks"""
+    for object in block_objects:
+        set_rc_values_for_block_from_pvs(object,pvs)
```
Comment (samueljackson92) [inline get_block_details.py]: `object` is a type in python. We should use a different name to prevent shadowing,

File: get_block_details.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+
+def add_rc_values(pvs, blocks):
+    """Look for RC values in the supplied list o pvs and assign them to the correct block."""
```
Comment (samueljackson92) [inline get_block_details.py]: This appear to be unimplemented. It should just be removed.

File: get_block_details.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+                if len(block_split) < 4:
+                    # Missing a value
+                    continue
+                else:
+                    value = block_split[value_index]
+                    alarm = block_split[alarm_index]
+
+            name = shorten_title(titles[i])
+            status = status_text[i]
+            blocks[name] = Block(name, status, value, alarm, True)
```
Comment (samueljackson92) [inline get_block_details.py]: This try block is huge. We should consider a new function. `parse_block` or `create_block` or something.

File: get_block_details.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+        except Exception as e:
+            logging.error("Unable to decode " + str(titles[i]))
+
+    return blocks
+
+# ans = get_info("http://localhost:4812/group?name=INST")
+# for a in ans:
+#     print a
```
Comment (samueljackson92) [inline get_block_details.py]: Commented code can be removed.

File: get_webpage.py [hunk 9]
```diff
@@ -21,8 +21,22 @@ def shorten_title(title):
...
+        title_list.append(title[-3])
+        title_list.append(rc_name)
+        for name in value_name:
+            if name in title:
+                title_list.append(name)
+                title = ':'.join(title_list)
+                return title
+    else:
+        title = title[-1]
+        return title
```
Comment (samueljackson92) [inline get_webpage.py]: This code looks identical to that in `shorten_rc_title`. We should refactor to just use a common function instead.

File: get_webpage.py [hunk 10]
```diff
@@ -147,6 +184,7 @@ def get_instpvs(url):
 
     """
     wanted = dict()
+    rc = dict()
```
Comment (samueljackson92) [inline get_webpage.py]: Why not use `{}`? It's shorter, more pythonic, and [it's faster...](https://stackoverflow.com/questions/5790860/and-vs-list-and-dict-which-is-better)

File: get_webpage.py [hunk 11]
```diff
@@ -155,6 +193,15 @@ def get_instpvs(url):
                     "TOTALCOUNTS", "DAETIMINGSOURCE", "MONITORCOUNTS", "MONITORSPECTRUM", "MONITORFROM", "MONITORTO",
                     "NUMTIMECHANNELS", "NUMSPECTRA"]
 
+    #rc_pvs = []
+    print blocks_all.values()[0]
```
Comment (samueljackson92) [inline get_webpage.py]: Print statements should be removed or modified into logger calls. I'm guessing these can just be removed.

File: get_webpage.py [hunk 12]
```diff
@@ -155,6 +193,15 @@ def get_instpvs(url):
...
+
+    try:
+        set_rc_values_for_blocks(blocks_all.values(), ans)
+    except Exception as e:
+        logging.error("Error in setting rc values for blocks: " + str(e))
+
+    print blocks_all.values()[0]
```
Comment (samueljackson92) [inline get_webpage.py]: As above.

File: webserver_tests.py [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+        test_block = "NEW_BLOCK"
+        test_pvs = {"NEW_BLOCK:RC:LOW.VAL" : 10, "MORGAN:RC:LOW.VAL" : 100,  "OLD_BLOCK:RC:LOW.VAL" : 7}
+
+        # Act
+        rc_values = get_rc_values_for_block(test_block, test_pvs)
+
+        # Assert
+        self.assertEquals(rc_values, 10)
+
+    def test_get_one_rc_value_for_block(self):
```
Comment (samueljackson92) [inline webserver_tests.py]: This is a redefinition of the test above, which means:
 - This test does not run
 - This test actually fails (see comment [above](https://github.com/ISISComputingGroup/JSON_bourne/pull/14/files#diff-516f5df5fa600fbbd033e61c35d81f94R81)) but when you run the tests they will not complain.

File: webserver_tests.py [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+        test_block = "NEW_BLOCK"
+        test_pvs = {"NEW_BLOCK:RC:LOW.VAL": 10, "NOT_NEW_BLOCK:RC:LOW.VAL": 100, "OLD_BLOCK:RC:LOW.VAL": 7}
+
+        # Act
+        rc_values = get_rc_values_for_block(test_block, test_pvs)
+
+        # Assert
+        self.assertEquals(rc_values, 10)
+
+    def test_get_one_rc_value_for_block(self):
```
Comment (samueljackson92) [inline webserver_tests.py]: As above.

File: webserver_tests.py [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+                    "NOT_NEW_BLOCK:RC:LOW.VAL": Block("", "", 100, "", ""),
+                    "OLD_BLOCK:RC:LOW.VAL": Block("", "", 7, "", "")}
+
+        # Act
+        rc_values = get_rc_values_for_block_from_object(test_block, test_pvs)
+
+        # Assert
+        self.assertEquals(rc_values, 10)
+
+    def test_get_one_rc_value_for_block(self):
```
Comment (samueljackson92) [inline webserver_tests.py]: As above.

File: webserver_tests.py [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+
+        # Act
+        description = test_block.get_description()
+
+        # Assert
+        self.assertEquals(description["rc_low"], 0)
+        self.assertEquals(description["rc_high"], 100)
+        self.assertEquals(description["rc_inrange"], False)
+
+    def test_when_rc_values_given_block_description_contains_rc_values(self):
```
Comment (samueljackson92) [inline webserver_tests.py]: Also a duplicate name.

File: get_block_details.py [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+def set_rc_values_for_block_from_pvs(block_object, pvs):
+    """Search pvs for RC values for given block and return them"""
+    block_name = block_object.get_name()
+    items = pvs.items()
+
+    for k, v in items:
+        test_block_name = k.split(':')[0]
+        if block_name == test_block_name:
+            if "LOW.VAL" == k.split(':')[-1]:
+                low_rc = block_object.set_rc_low(v.get_value())
```
Comment (samueljackson92) [inline get_block_details.py]: This looks wrong? shouldn't we be calling `get_rc_inrange` instead of `set_rc_inrange`?

**Comment by samueljackson92** [comment]
@Tom-Willemsen and I found a bug with the block server that prevents the instrument status from correctly being written to the archive. The fix looks simple so we decided to do it here as it's important that the information shown on dataweb is consistent with the Ibex gui.

**Comment by samueljackson92** [comment]
This PR has uncovered a deeper problem with the archiver for the INST group. See issue ISISComputingGroup/IBEX#2446. For now, we decided the best course of action is to submit the work in this PR but use a hard coded flag to make sure it is not served to the user by default.

For testing purposes you can turn on the new functionality by editing the `RETURN_RC_VALUES` value at the top of `block.py`. When issue ISISComputingGroup/IBEX#2446 has been resolved this flag can be removed.

File: display_blocks.js (hunk 1 above)
Comment (DominicOram) [inline display_blocks.js]: Yes, this should be reverted, I will do this as part of reviewing.

File: block_utils_tests.py [hunk 18]
```diff
@@ -0,0 +1,380 @@
...
+        #Arrange
+        test_pv = "TE:NDLT910:CS:SB:NEW_BLOCK:RC:INRANGE.VAL"
+
+        #Act
+        shortened_title = shorten_title(test_pv)
+
+        #Assert
+        self.assertEquals(shortened_title, "NEW_BLOCK:RC:INRANGE.VAL")
+
+    def test_shorten_title_with_rc_enabled(self):
```
Comment (DominicOram) [inline block_utils_tests.py]: This test does not pass

**Comment by DominicOram** [comment]
Can you put the unit tests in a separate folder. Also, I can't seem to find RETURN_RC_VALUES that you mentioned? As far as I can see it's always on?

**Comment by samueljackson92** [comment]
@DominicOram I missed some commits. Sorry :man_facepalming:

**Comment by DominicOram** [comment]
No worries, all good now.

**Pull request merged**